### PR TITLE
Add `oauth2_scopes` with constants and functions for templating scopes

### PR DESCRIPTION
**Pull request by rudyardrichter**
- Add `oauth2_scopes.py` with a bunch of scope constants as well as some simple functions for templating scopes as URNs or URLs.

I didn't want to get clever with dynamically assigning attributes on the `XScopes` objects. One might argue for all caps: `XScopes.SOME_SCOPE`.

Does the `XScopes.some_scope` interface seem useful and easy to follow?

### CODE REVIEW

**Comment by rudyardrichter** [comment]
Also, this is pretty much all the random scope constants I could scrape from a pass through our repositories, but I could possibly have missed some. I'm not sure whether including Automate in the constants makes sense as the bulk of the scopes used are generated dynamically for each flow, but we could add action providers like `"https://auth.globus.org/scopes/actions.globus.org/hello_world"`.

**Review by sirosen — COMMENTED** [review]
Sorry for the long delay in getting review done.
There are some minor changes needed, but we should also follow up on a couple of these scopes -- as noted below -- to make sure they are good to include.

---

The `<X>Scopes.<scope_name>` interface is pretty easy to use, and should be easy to document. It might be good to add a page to the docs for this, to ensure that these functions and constants are included. We can do that in a follow-up if you prefer.

One thing which is missing is that we need to decide on where these imports should be done from for users. Some options:
- `src/globus_sdk/__init__.py` --> `from globus_sdk import urn_scope_string`
- `src/globus_sdk/scopes.py` --> `from globus_sdk.scopes import urn_scope_string`

I think I prefer the latter, but both are fine.

> I didn't want to get clever with dynamically assigning attributes on the `XScopes` objects.

I think keeping it simple for a first cut is good, but I did raise the possibility of a `ScopeBuilder` or `ResourceServerScopes` object below. My reasoning centers more around dynamic cases, like GCS scopes, for which some kind of object to represent "the scopes for a resource server" might make sense as a model.

Another way to look at it is that if the interface for `AuthScopes` and `TransferScopes` is good, it may be nice to allow users to build a similar object dynamically, as in `MyGCSScopes = Foo(<id>)`.

File: src/globus_sdk/services/auth/oauth2_scopes.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    """
+    Return the data access scope string for the given client.
+
+    :param client_id: The UUID for an auth client.
+    :type client_id: str
+    """
+    return url_scope_string(client_id, "data_access")
+
+
+class AuthScopes:
```
[the remark this answers is not in the document]
Comment (sirosen) [reply src/globus_sdk/services/auth/oauth2_scopes.py:50]: After our 2 minutes of discussion, I think I am convinced that strings are better.

File: src/globus_sdk/services/auth/oauth2_scopes.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+class GroupsScopes:
+    all = "urn:globus:auth:scope:groups.api.globus.org:all"
+    view_my_groups_and_memberships = (
+        "urn:globus:auth:scope:groups.api.globus.org:view_my_groups_and_memberships"
+    )
+
+
+class NexusScopes:
+    credential_cache = "urn:globus:auth:scope:nexus.api.globus.org:credential_cache"
```
[the remark this answers is not in the document]
Comment (sirosen) [reply src/globus_sdk/services/auth/oauth2_scopes.py:72]: @aaschaer, can you check Transfer for this and come back to tell us the result here? Thanks much.

**Comment by rudyardrichter** [comment]
Per discussion with @sirosen: may want to attach to service clients, so `TransferClient.scopes` has relevant scopes attached, and `tc.scopes.all` gives you the "transfer-all" scope. If not in this PR, then in some follow-up soon.

File: src/globus_sdk/services/auth/oauth2_scopes.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (aaschaer) [reply src/globus_sdk/services/auth/oauth2_scopes.py:72]: Yes, credential_cache is Transfer's scope for calling Nexus, we should remove it from here

File: src/globus_sdk/services/auth/oauth2_scopes.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    email = "email"
+    profile = "profile"
+    view_authentications = "urn:globus:auth:scope:auth.globus.org:view_authentications"
+    view_clients = "urn:globus:auth:scope:auth.globus.org:view_clients"
+    view_clients_and_scopes = (
+        "urn:globus:auth:scope:auth.globus.org:view_clients_and_scopes"
+    )
+    view_identities = "urn:globus:auth:scope:auth.globus.org:view_identities"
+    view_identity_set = "urn:globus:auth:scope:auth.globus.org:view_identity_set"
+    view_ssh_public_keys = "urn:globus:auth:scope:auth.globus.org:view_ssh_public_keys"
```
[the remark this answers is not in the document]
Comment (sirosen) [reply src/globus_sdk/services/auth/oauth2_scopes.py:61]: Judging from what we've learned talking with Auth, I think we should remove this. Seems to me like it's not really usable.

**Comment by sirosen** [comment]
I think the revised version of this is spot-on. I can't think of any changes I would make to it -- only future things we might add on top of it.

**Pull request merged**